### PR TITLE
Rename `-Xlint:named-booleans` to `-Wunnamed-boolean-literal` (and no longer include it in `-Xlint`)

### PR DESCRIPTION
**Pull request by som-snytt**
The existing lint to check that boolean literal arguments in method calls use named arg syntax (`f(flag = true)`) is moved to a warning option `-Wunnamed-boolean-literal`.

The warning is issued if the unnamed arg may be inadvertently swapped with another: there is another unnamed boolean arg or another boolean parameter that is supplied using a default argument.

A stricter warning is available under `-Wunnamed-boolean-literal-strict`, which warns about any unnamed boolean arg, except for the pattern of an `assert` method that takes exactly one boolean parameter in first position.

The exclusion for `assert` assumes that such a method will have a name that conveys how the boolean arg is used, so that naming the arg is superfluous.

This warning is on a par with `-Wvalue-discard`, which is also an ordinary language feature, but may be deemed warnable for some projects.

Follow-up to https://github.com/scala/scala/pull/10612

Removes the exclusion Lukas doubted at https://github.com/scala/scala/pull/10612#discussion_r1420102963

### CODE REVIEW

**Comment by SethTisue** [comment]
I've posted a call for feedback on this in various places (including the Known issues section of the 2.13.13 release notes), with a link to this PR.

**Comment by He-Pin** [comment]
In pekko we just turn it off for now, otherwise, there are many code that need to be changed, but overall I think this is a good feature.
`
-Wconf:cat=unused-nowarn:s,cat=lint-named-booleans:s,cat=other-shadowing:s,any:e
`

File: test/files/neg/named-booleans-relaxed.check [hunk 1]
```diff
@@ -0,0 +1,21 @@
+named-booleans-relaxed.scala:22: warning: Boolean literals should be passed using named argument syntax for parameter x.
+  val x0 = c.f(17, true, false) // warn
```
Comment (He-Pin) [inline test/files/neg/named-booleans-relaxed.check:2]: Add a case for a single `boolean` parameter, which expected no warning.

Comment (som-snytt) [reply test/files/neg/named-booleans-relaxed.check:2]: `c.check(true, "OK")` but it also does not warn in "strict" mode because of Lukas rules.

File: src/compiler/scala/tools/nsc/settings/Warnings.scala [hunk 2]
```diff
@@ -120,6 +120,7 @@ trait Warnings {
   val warnValueDiscard     = BooleanSetting("-Wvalue-discard", "Warn when non-Unit expression results are unused.") withAbbreviation "-Ywarn-value-discard"
   val warnNumericWiden     = BooleanSetting("-Wnumeric-widen", "Warn when numerics are widened.") withAbbreviation "-Ywarn-numeric-widen"
   val warnOctalLiteral     = BooleanSetting("-Woctal-literal", "Warn on obsolete octal syntax.") withAbbreviation "-Ywarn-octal-literal"
+  val warnNamedLiteral     = BooleanSetting("-Wnamed-literal", "Warn name literals in an application that may be swapped, unless param is @deprecatedName.")
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/settings/Warnings.scala]:
```suggestion
  val warnNamedLiteral     = BooleanSetting("-Wnamed-literal", "Warn when unnamed literal arguments may be swapped, unless param is @deprecatedName.")
```

File: src/compiler/scala/tools/nsc/settings/ScalaSettings.scala [hunk 3]
```diff
@@ -216,6 +216,7 @@ trait ScalaSettings extends StandardScalaSettings with Warnings { _: MutableSett
    * -Y "Private" settings
    */
   val Yhelp           = BooleanSetting    ("-Y", "Print a synopsis of private options.")
+  val Ylint           = BooleanSetting    ("-Ylint", "Use strict lints.")
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/settings/ScalaSettings.scala]: I'm skeptical about adding that flag.

File: src/compiler/scala/tools/nsc/typechecker/RefChecks.scala [hunk 4]
```diff
@@ -1724,26 +1724,32 @@ abstract class RefChecks extends Transform {
...
-          })
+        if (!settings.Ylint.value || !isAssertParadigm(params)) {
+          val suspicious = args.lazyZip(params).iterator
+            .filter {
+              case (arg @ Literal(Constant(_: Boolean)) , param) =>
+                arg.hasAttachment[UnnamedArg.type] && param.tpe.typeSymbol == BooleanClass && !param.deprecatedParamName.contains(nme.NO_NAME)
+              case _ => false
+            }
+            .toList
+          val warn = if (settings.Ylint.value) suspicious.nonEmpty else suspicious.lengthCompare(2) >= 0
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/RefChecks.scala]: Maybe we should also warn if there's one literal arg but the function has multiple boolean parameters

```scala
scala> def f(x: Int, y: Boolean = true, z: Boolean = true) = 0
def f(x: Int, y: Boolean, z: Boolean): Int

scala> f(1, y = true, false) // warn?
val res12: Int = 0

scala> f(1, true) // warn?
val res13: Int = 0
```

Otherwise warning only for multiple literals seems a good way to tone it down.

**Review by lrytz — COMMENTED** [review]
Where were the complaints? I don't have the context.

Is demoting to `-W` still necessary with the adjustments in this PR?

**Comment by milessabin** [comment]
Scala 2 appears to accept unnamed arguments after named, which means that a quick fix simply inserting the argument name is possible.  I was surprised to find this was accepted, because, IMO, this is a _bug_ in Scala 2, long standing presumably, which has just been fixed in Scala 3: https://github.com/lampepfl/dotty/pull/18363.

This means that the simple quick fix of inserting the name will risk breaking cross builds with Scala 3. A more comprehensive fix which moves named boolean argument to the end of arglists might be possible (evaluation order effects?), but a lot more complicated.

Ultimately I think an implication of this lint is that library authors should move all booleans to the end of parameter lists to allow downstream users to enable the lint, or not, with a minimum of fuss. But this impact on APIs seems wildly excessive to me.

**Comment by lrytz** [comment]
> Scala 2 appears to accept unnamed arguments after named

Only if all preceding named arguments are on the correct position. I beleive Scala 3 behaves the same, also after https://github.com/lampepfl/dotty/pull/18363. IIRC Scala 3 used to allow *more* unnamed after named than Scala 2, and that PR brought the two in line.

**Comment by milessabin** [comment]
> I beleive Scala 3 behaves the same, also after https://github.com/lampepfl/dotty/pull/18363

It does. I take it back.

File: src/compiler/scala/tools/nsc/settings/ScalaSettings.scala (hunk 3 above)
Comment (som-snytt) [reply src/compiler/scala/tools/nsc/settings/ScalaSettings.scala]: I was going to call it `-Ylrytz-rules`.

File: src/compiler/scala/tools/nsc/typechecker/RefChecks.scala (hunk 4 above)
Comment (som-snytt) [reply src/compiler/scala/tools/nsc/typechecker/RefChecks.scala]: Somewhere I commented that I thought what the user wrote at use site is more relevant than the definition. That is an untested claim.

Edit: on re-reading, I understand you mean: warn if the intended "target" param is ambiguous. That is obvious for res13, less obvious for res12. I like to write `@deprecated("avoid me", since="1.0")` for clarity, but I don't name both (because I don't remember if the name is `msg` or `message`, such is our penchant for abbreviating).

**Comment by som-snytt** [comment]
The misleading text in the tour has been updated at https://docs.scala-lang.org/tour/named-arguments.html

**Comment by som-snytt** [comment]
As lrytz mentioned on the PR, the two aspects are safety (guard against swapping) and style (readability).

This commit puts safety under `-Wnamed-literal` and style under `-Ylint` (name is arbitrary for private option).

The `-W` option suggests that in future, other literals may wish to be linted too.

I think trying to find a "sweet spot" for all use cases is a fool's errand. Personally, I don't want to see `assert(false)` but `fail()`. All the more so with `f(true)`. I would compile with `-W:11`, at least in CI.

File: src/compiler/scala/tools/nsc/typechecker/RefChecks.scala (hunk 4 above)
Comment (som-snytt) [reply src/compiler/scala/tools/nsc/typechecker/RefChecks.scala]: The paradigm here is to consider "effectively unnamed boolean args", which are unnamed literals and default args. The relaxed check warns if there are at least 2 effectively unnamed args, at least one of which is a literal.

**Comment by som-snytt** [comment]
Rebased and took lrytz's two suggestions. Copying my inline comment:

The paradigm is to consider "effectively unnamed boolean args", which are unnamed literals and default args. The relaxed check warns if there are at least 2 effectively unnamed args, at least one of which is a literal.

There is an old paulp comment to the effect that a boolean arg with a default is peak evil. I did not add that to the warning text, however.

Edit: also realized (again) that I can't enable `-Ylinternal` for its intended internal purpose until restarr.

**Comment by som-snytt** [comment]
More random scripted test errors.
```
[info] [error] ## Exception when compiling 1 sources to /tmp/sbt_3bd3d6f3/target/classes
[info] [error] java.lang.NoClassDefFoundError: scala/unchecked
[info] [error] scala.reflect.internal.Definitions$DefinitionsClass.UncheckedClass$lzycompute(Definitions.scala:1357)
[info] [error] scala.reflect.internal.Definitions$DefinitionsClass.UncheckedClass(Definitions.scala:1357)
[info] [error] scala.tools.nsc.typechecker.Checkable.scala$tools$nsc$typechecker$Checkable$$uncheckedOk(Checkable.scala:131)
[info] [error] scala.tools.nsc.typechecker.Checkable$InferCheckable.checkCheckable(Checkable.scala:309)
[info] [error] scala.tools.nsc.typechecker.Checkable$InferCheckable.checkCheckable$(Checkable.scala:309)
[info] [error] scala.tools.nsc.typechecker.Infer$Inferencer.checkCheckable(Infer.scala:210)
[info] [error] scala.tools.nsc.typechecker.Typers$Typer.typedTypeApply(Typers.scala:4521)

[snip]

[info] [info] welcome to sbt 1.9.9 (Oracle Corporation Java 1.8.0_332)
[info] [info] loading project definition from /tmp/sbt_8b7edbef/project
[info] [info] loading settings for project sbt_8b7edbef from common.sbt ...
[info] [info] set current project to sbt_8b7edbef (in build file:/tmp/sbt_8b7edbef/)
[info] [info] compiling 3 Scala sources to /tmp/sbt_8b7edbef/target/classes ...
[info] [error] ## Exception when compiling 3 sources to /tmp/sbt_8b7edbef/target/classes
[info] [error] java.lang.NoClassDefFoundError: scala/collection/ArrayOps$ArrayView
[info] [error] scala.tools.nsc.Reporting$WConf$.$anonfun$parse$1(Reporting.scala:861)
[info] [error] scala.tools.nsc.Reporting$WConf$.parse(Reporting.scala:859)
[info] [error] scala.tools.nsc.Reporting$PerRunReporting.wconf$lzycompute(Reporting.scala:51)
[info] [error] scala.tools.nsc.Reporting$PerRunReporting.wconf(Reporting.scala:51)
[info] [error] scala.tools.nsc.Reporting$PerRunReporting.scala$tools$nsc$Reporting$PerRunReporting$$issueWarning(Reporting.scala:189)
[info] [error] scala.tools.nsc.Reporting$PerRunReporting.issueIfNotSuppressed(Reporting.scala:234)
[info] [error] scala.tools.nsc.Reporting$PerRunReporting.deprecationWarning(Reporting.scala:300)
[info] [error] scala.tools.nsc.Reporting$PerRunReporting.deprecationWarning(Reporting.scala:306)
[info] [error] scala.tools.nsc.Reporting$PerRunReporting.deprecationWarning(Reporting.scala:312)
[info] [error] scala.tools.nsc.typechecker.RefChecks$RefCheckTransformer.scala$tools$nsc$typechecker$RefChecks$RefCheckTransformer$$checkUndesiredProperties(RefChecks.scala:1330)
[info] [error] scala.tools.nsc.typechecker.RefChecks$RefCheckTransformer.transformSelect(RefChecks.scala:1771)

```
etc. /cc @lrytz

File: src/compiler/scala/tools/nsc/typechecker/RefChecks.scala [hunk 5]
```diff
@@ -1715,35 +1715,52 @@ abstract class RefChecks extends Transform {
...
+                  }
+                case _ => acc
+              }
+            }
+          val warn = !unnamed.isEmpty && (settings.Ylint.value || numSuspicious >= 2)
+          if (warn)
+            unnamed.reverse.foreach {
+              case (arg, param) =>
+                val msg = s"Boolean literals should be passed using named argument syntax for parameter ${param.name}."
+                val action = runReporting.codeAction("name boolean literal", arg.pos.withEnd(arg.pos.start), s"${param.name} = ", msg)
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/RefChecks.scala]: `arg.pos.focusStart`

File: src/compiler/scala/tools/nsc/typechecker/RefChecks.scala [hunk 6]
```diff
@@ -1715,35 +1715,52 @@ abstract class RefChecks extends Transform {
...
+        if (checkable) {
+          val (unnamed, numSuspicious) = args.lazyZip(params).iterator
+            .foldLeft((List.empty[(Tree, Symbol)], 0)) { (acc, ap) =>
+              ap match {
+                case (arg, param) if param.tpe.typeSymbol == BooleanClass && !param.deprecatedParamName.contains(nme.NO_NAME) =>
+                  arg match {
+                    case Literal(Constant(_: Boolean)) =>
+                      if (isUnnamedArg(arg)) (ap :: acc._1, acc._2 + 1) else acc
+                    case _ =>
+                      if (arg.symbol != null && arg.symbol.isDefaultGetter) (acc._1, acc._2 + 1) else acc
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/RefChecks.scala]: `isDefaultGetter` doesn't work when the named args tranformation did its thing. Maybe an unlikely corner case though.

```scala
scala> def bul = true
def bul: Boolean

scala> def f(n: Int, inv: Boolean = false, up: Boolean = false, down: Boolean = false) = 0
def f(n: Int, inv: Boolean, up: Boolean, down: Boolean): Int

scala> f(1, true, up = bul)
            ^
       warning: Boolean literals should be passed using named argument syntax for parameter inv. [quickfixable]
val res0: Int = 0

scala> f(1, true, down = bul)
val res1: Int = 0
```

```scala
      private[this] val res2: Int = {
        <artifact> val x$1: Int(1) = 1;
        <artifact> val x$2: Boolean(true) = true;
        <artifact> val x$3: Boolean = $read.this.$line3$read.$iw.bul;
        <artifact> val x$4: Boolean = $read.this.$line4$read.$iw.f$default$3;
        $read.this.$line4$read.$iw.f(1, true, x$4, x$3)
      };
```

**Review by lrytz — COMMENTED** [review]
Looks good in general, but I'm still not in favor of the -Ylinterator flag. Who is it intended for?

**Comment by lrytz** [comment]
> random scripted test errors.

🤷‍♂️ no idea... clicked rebuild.

File: src/compiler/scala/tools/nsc/typechecker/RefChecks.scala (hunk 6 above)
Comment (som-snytt) [reply src/compiler/scala/tools/nsc/typechecker/RefChecks.scala]: The unlikely corner case complains the loudest.

**Comment by som-snytt** [comment]
I'll rename to `-Ylinterator`. I should have asked you for a good name first. That audience is internal us. I think the single boolean literal arg is odoriferous enough to warrant a private option. If people try it out and demand it, it could be promoted.

The broad lesson, as well, is to introduce noise under a flag and then promote to lint under public pressure, rather than the other way around.

File: src/compiler/scala/tools/nsc/typechecker/RefChecks.scala (hunk 5 above)
Comment (som-snytt) [reply src/compiler/scala/tools/nsc/typechecker/RefChecks.scala]: That is not a range position, but I suppose it doesn't matter? or could it matter? Would have to refresh memory on position invariants, transparency, how edits are applied, etc.

**Comment by som-snytt** [comment]
Why did applications with named args stop working?

They were out of order.

File: src/compiler/scala/tools/nsc/typechecker/RefChecks.scala (hunk 6 above)
Comment (som-snytt) [reply src/compiler/scala/tools/nsc/typechecker/RefChecks.scala]: thanks, this is another instance of nothing good happens after midnight. I was looking at this, but a dumb bug in my refactor used up the rest of my midnight oil. It's an edge case because it requires many boolean parameters and one boolean literal arg.

It would be nice if default arg were an attachment instead of a name test. A name can be long and linting prefers efficient up-front tests even if emitting diagnostics is expensive.

**Comment by som-snytt** [comment]
`NamedApplyInfo` now carries the original `Apply` and besides the `Block` for the rewrite, attaches to the result `Apply`.

Args in the result `Apply` retain attachment~s~ (`Unnamed`) from original args. (Note: don't copy all attachments, as `MacroExpansion` is used by linting.)

**Comment by som-snytt** [comment]
Intended to rename the `-Y` to `-Wnamed-boolean` as a narrower version of `-Wnamed-literal` (though the latter is still only about booleans).

Edit: done. Can be squashed.

**Comment by SethTisue** [comment]
I think I would expect the name to be `-Wunnamed-boolean` rather than the reverse? wdyt?

**Comment by som-snytt** [comment]
That would be a warning about mononyms, as it's pronounced "one-named boolean".

I considered changing it, but ran out of options.

On reconsideration, I see the issue is that the word is not "unnamed" but "positional". `-Wpositional-boolean-literal-arg`.

I'll keep the phone lines open right up until squashing, for any further suggestions.

**Comment by SethTisue** [comment]
(I'm pretty swamped this week and next and will probably be slow with the more in-depth feedback, I'm afraid.)

File: src/compiler/scala/tools/nsc/settings/Warnings.scala [hunk 7]
```diff
@@ -121,6 +121,8 @@ trait Warnings {
   val warnValueDiscard     = BooleanSetting("-Wvalue-discard", "Warn when non-Unit expression results are unused.") withAbbreviation "-Ywarn-value-discard"
   val warnNumericWiden     = BooleanSetting("-Wnumeric-widen", "Warn when numerics are widened.") withAbbreviation "-Ywarn-numeric-widen"
   val warnOctalLiteral     = BooleanSetting("-Woctal-literal", "Warn on obsolete octal syntax.") withAbbreviation "-Ywarn-octal-literal"
+  val warnNamedLiteral     = BooleanSetting("-Wnamed-literal", "Warn when unnamed literal arguments may be swapped, unless parameter has @deprecatedName.")
+  val warnNamedBoolean     = BooleanSetting("-Wnamed-boolean", "Boolean literal arguments should be named.").enabling(warnNamedLiteral :: Nil)
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/settings/Warnings.scala]: OK, two `-W` flags is fine with me, certainly better than `-Ylint`.

I'd adjust the flag names, for the existing ones `-Wx` warns about doing x, not about "not having done x". `unnamed-litrals` and `unnamed-boolean` maybe.

Do you think extending the first warning to other primitive types would cause issues? Not needed for this PR, but since the flag has a general name we should give it a some thought.

**Review by lrytz — COMMENTED** [review]
Looks good!

File: src/compiler/scala/tools/nsc/settings/Warnings.scala (hunk 7 above)
Comment (lrytz) [reply src/compiler/scala/tools/nsc/settings/Warnings.scala]: I think extending it to other types might not be as obvious, there are more conventions
  - `Point(3, 7)` // obviously x, y
  - `Complex(3.3, 1.8)` // re, im of course

Comment (som-snytt) [reply src/compiler/scala/tools/nsc/settings/Warnings.scala]: `@deprecatedName` works, except that it also discourages naming. Maybe in future, a different annotation just for this case.

**Comment by som-snytt** [comment]
You're supposed to try on the shoes and walk about the store a bit.

On one foot is `-Wliteral-args` (which warns if two args might be swapped, currently only boolean literals) and on the other foot is `-Wboolean-literal` (which warns about a boolean literal arg that is positional, modulo the lrytz exclusion).

I don't think my toe is touching in front.

But it drives me bonkers that github line-splits `-` and `Wboolean-literal` in backticks. I added words just for formatting.

**Comment by som-snytt** [comment]
Forgot to update the junit test.

File: src/compiler/scala/tools/nsc/settings/Warnings.scala [hunk 8]
```diff
@@ -121,6 +121,8 @@ trait Warnings {
   val warnValueDiscard     = BooleanSetting("-Wvalue-discard", "Warn when non-Unit expression results are unused.") withAbbreviation "-Ywarn-value-discard"
   val warnNumericWiden     = BooleanSetting("-Wnumeric-widen", "Warn when numerics are widened.") withAbbreviation "-Ywarn-numeric-widen"
   val warnOctalLiteral     = BooleanSetting("-Woctal-literal", "Warn on obsolete octal syntax.") withAbbreviation "-Ywarn-octal-literal"
+  val warnNamedLiteral     = BooleanSetting("-Wliteral-args", "Warn when positional literal arguments should be named, unless parameter has @deprecatedName.")
+  val warnNamedBoolean     = BooleanSetting("-Wboolean-literal", "Boolean literal arguments should be named.").enabling(warnNamedLiteral :: Nil)
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/settings/Warnings.scala]:
```suggestion
  val warnNamedLiteral     = BooleanSetting("-Wunnamed-boolean-literal", "Warn about unnamed boolean literals if there is more than one or defaults are used, unless parameter has @deprecatedName.")
  val warnNamedBoolean     = BooleanSetting("-Wunnamed-boolean-literal-strict", "Warn about all unnamed boolean literals, unless parameter has @deprecatedName or the method has a single leading boolean parameter").enabling(warnNamedLiteral :: Nil)
```

**Review by lrytz — COMMENTED** [review]
Re-reviewed. The implementation LGTM, I'd change the flag names. My suggestion is rather verbose but I think easier to understand.

Leaving to @SethTisue for 👓

**Comment by som-snytt** [comment]
I don't have any fight left in me over names, so I will take the suggestion.

File: src/compiler/scala/tools/nsc/typechecker/RefChecks.scala [hunk 9]
```diff
@@ -1703,47 +1703,75 @@ abstract class RefChecks extends Transform {
...
-      }
-      if (settings.lintNamedBooleans && !sym.isJavaDefined && !args.isEmpty) {
+      if (settings.warnNamedLiteral.value && !sym.isJavaDefined && !args.isEmpty) {
+        def isUnnamedArg(t: Tree) = t.hasAttachment[UnnamedArg.type]
+        def isDefaultArg(t: Tree) = t.symbol != null && (
+          t.symbol.isDefaultGetter || t.symbol.name.startsWith(nme.NAMEDARG_PREFIX) && {
+            analyzer.NamedApplyBlock.namedApplyInfo(currentApplication) match {
+              case Some(analyzer.NamedApplyInfo(_, _, _, _, original)) =>
+                val treeInfo.Applied(_, _, argss) = original
+                val orig = argss.head(t.symbol.name.decoded.stripPrefix(nme.NAMEDARG_PREFIX).toInt - 1)
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/typechecker/RefChecks.scala]: this is totally wrong. `x$42` is just a fresh name, not an index into the param list.

**Comment by som-snytt** [comment]
Someday, instead of coding, we'll just ask the A.I. to "rebuild" and maybe "tweak". The second commit fixes detecting "a default arg was used for a boolean param in a named arg block". `$default$7` means param 7. We just need to count them, as a threshold for warning (for the non-strict option). Added the test drawn from "real code", if reflect is real code.

A third commit adds a check for `f$default$7$extension`, which is not seen at refchecks but the utility method was decommissioned in https://github.com/scala/scala/commit/4ed31f48515de8a3f8a3b2120c6c7ad5fca77889 for "value classes in REPL". I'm not sure why unused check was running late, but the previous usage was to peel off the method name; here, it is to peel off the parameter index.

Updated the PR OP comment @SethTisue 

Sorry @lrytz for the re-review re-request.

**Review by SethTisue — APPROVED** [review]
I agree on the new naming here.

@lrytz you can merge after taking the last look it seems Som wants you to take

**Review by lrytz — APPROVED** [review]
All good, thanks for fixing the defaults!
